File: 2022/13/solution.pas
{ solution.pas -- distress signal -- Troy Brumley BlameTroi@gmail.com }

program solution;

{ expects to be compiled with the following two mode flags and command
  line/fpc.cfg options of -Ciort -Sgix -Sh- for full checks, allowing
  exceptions, goto, inlining, and prefer classic strings. }
// {$LONGSTRINGS OFF}            { to be sure, use AnsiString or pchar explicitly }
{$BOOLEVAL ON}                { standard pascal expects this }

{ advent of code 2022 day 13 -- distress signal

  for part one, identify communication packets that are correctly
  ordered. the packets are lists and this looks rather lispish in
  lean, but i'll stick with pascal. }

{$LONGSTRINGS ON } { <-------- the input is very wide }

const
   MAXPAIRS = 500;

type
   integer = int64;            { by default ints are 64 bits this decade}


{$I aocinput.p }
{$I str2int.p }
{$I int2str.p }


{ the only way i've been able to get an algorithm that doesn't offend
  my sensibilities is to get closer and closer to a lisp list. i guess
  they knew what they were doing way back then. a list is at least two
  nodes. the last node is always empty. as the problem for advent of
  code is dealing with lists of integers, the value payload is just an
  integer. }


type
   patom = ^tatom;
   tatom = record
              isnull : integer;   { 0 = false, 1 = true }
              value  : integer;
           end;

   plist = ^tlist;
   tlist = record
              atom   : patom;
              nested : plist;
              next   : plist;
           end;


{ list node manipulators }

{ create a new list ground/nil node }

function newempty : plist;
begin
   new(newempty);
   with newempty^ do begin
      atom := nil;
      nested := nil;
      next := nil;
   end;
end;


{ a list can contain another list }

procedure setnested(xs, ys : plist);
begin
   with xs^ do begin
      if atom <> nil then begin
         writeln(stderr, '*** assigning nested to atom node ***');
         halt;
      end;
      nested := ys;
   end;
end;


function getnested(xs : plist) : plist;
begin
   getnested := xs^.nested;
end;


{ a list can contain an atomic value, in this application an integer }

procedure setvalue(xs : plist; n : integer);
begin
   with xs^ do begin
      if nested <> nil then begin
         writeln(stderr, '*** assigning value to nesting node ***');
         halt;
      end;
      { allocate storage for the atom, or reuse if one already there }
      if atom = nil then
         new(atom);
      atom^.isnull := 0;
      atom^.value := n;
   end;
end;


function getvalue(xs : plist) : integer;
begin
   getvalue := xs^.atom^.value;
end;


{ list nodes chain together for form the actual list }

procedure setnext(xs, ys : plist);
begin
   xs^.next := ys;
end;

function addnewempty(xs : plist) : plist;
begin
   xs^.next := newempty;
   addnewempty := xs^.next;
end;


{ various list node predicates }

function isempty(x : plist) : boolean; inline;
begin
   isempty := (x^.atom = nil) and (x^.nested = nil);
end;

function hasvalue(x: plist) : boolean; inline;
begin
   hasvalue := x^.atom <> nil;
end;

function hasnested(x : plist) : boolean; inline;
begin
   hasnested := x^.nested <> nil;
end;

function hasnext(x : plist) : boolean; inline;
begin
   hasnext := x^.next <> nil;
end;

function isnilorempty(x : plist) : boolean; inline;
begin
   if x = nil then
      isnilorempty := true
   else
      isnilorempty := isempty(x);
end;

function notnilorempty(x : plist) : boolean; inline;
begin
   notnilorempty := not isnilorempty(x);
end;


{ release the list any linked atom values and sub lists, clearing any
  storage before it is released. }

procedure freelist(xs : plist);

var
   nx : plist;

begin
   while xs <> nil do begin
      { hang on to next pointer }
      nx := xs^.next;

      { free chained lists }
      if hasnested(xs) then begin
         freelist(xs^.nested);
         xs^.nested := nil;
      end;

      { free value if there is one }
      if xs^.atom <> nil then begin
         xs^.atom^.isnull := 0;
         xs^.atom^.value := 0;
         dispose(xs^.atom);
         xs^.atom := nil;
      end;

      { release and advance }
      xs^.next := nil;
      dispose(xs);

      xs := nx;
   end;
end;

function listfromvalue(x : plist) : plist;
var
   y : plist;
begin
   y := newempty;
   if hasvalue(x) then
      setvalue(y, getvalue(x));
   listfromvalue := y;
end;

{ recursive helper for aslist. should be called with pos pointing at
  a list open bracket [ and will return as each list close bracket
  ] is reached. if everything is nested properly, it works. there is
  no attempt at input validation. }

function aslistr(    str : string;
                 var pos : integer) : plist;
var
   xs, ys : plist;
   n      : integer;

begin
   { guards and special cases }
   if length(str) < 2 then begin
      aslistr := nil;
      exit;
   end;

   if (str[pos] = '[') and (str[pos+1] = ']') then begin
      pos := pos + 2;
      aslistr := newempty;
      exit;
   end;

   if (str[pos] <> '[') or (str[length(str)] <> ']') then begin
      aslistr := nil;
      pos := pos + 1;
      exit;
   end;

   { first list node is also function result }
   xs := newempty;
   aslistr := xs;

   { consume opening [ }
   pos := pos + 1;

   while pos <= length(str) do begin

      { commas and spaces are white space }
      if (str[pos] = ' ') or (str[pos] = ',') then begin
         pos := pos + 1;
         continue;
      end;

      { end of the current list }
      if str[pos] = ']' then begin
         pos := pos + 1;
         break;
      end;

      { starting a nested list }
      if str[pos] = '[' then begin
         ys := aslistr(str, pos);
         setnested(xs, ys);
         xs := addnewempty(xs);
         continue;
      end;

      { add atom to list, only ints supported }
      if str[pos] in ['0'..'9'] then begin
         n := str2int(str, pos);
         setvalue(xs, n);
         xs := addnewempty(xs);
         continue;
      end;

      { error }
      writeln(stderr, 'error ', str[pos], ' at ', pos);
      pos := pos + 1;

   end;

end;


{ from a list of integers and nested lists of integers in string form,
  create a navigable set of nodes. the ugly details are in a recursive
  helper. }

function aslist(str : string) : plist;
var
   pos : integer;

begin
   pos := 1;
   aslist := aslistr(str, pos);
end;


{ given a list of nodes, return a string representation. }

function printlist(xs : plist) : string;

begin
   printlist := '[';
   while not isempty(xs) do begin               { end of (sub) list reached }
      if hasnested(xs) then begin               { sub list }
         printlist := printlist + printlist(getnested(xs));
         if not isempty(xs^.next) then
            printlist := printlist + ',';
         xs := xs^.next;
         continue;
      end;
      if hasvalue(xs) then begin                { or atom }
         printlist := printlist + int2str(getvalue(xs));
         if not isempty(xs^.next) then
            printlist := printlist + ',';
         xs := xs^.next;
         continue;
      end;
      { error }
      writeln(stderr, 'invalid list construction');
      break;
   end;
   printlist := printlist + ']';
end;


{ global storage for the aoc part of this }

type
   tpair =  record
               left    : plist;
               right   : plist;
               inorder : integer;
            end;

var
   pair       : array[1 .. MAXPAIRS] of tpair;
   packet     : array[1 .. MAXPAIRS*2] of plist;
   numpairs   : integer;
   numpackets : integer;


{ clear out the arrays for parts one and two }

procedure init;

begin
   for numpairs := 1 to MAXPAIRS do
      with pair[numpairs] do begin
         left := nil;
         right := nil;
         inorder := 0;
         packet[numpairs*2 - 1] := nil;
         packet[numpairs*2] := nil;
      end;
   numpairs := 0;
   numpackets := 0;
end;


{ compare two lists using the rules for part one and return via the
  negative zero positive convention.

  compare the lists item by item, with the following behaviors for
  nested lists:

  a longer list comes after an otherwise equivalent list

  if one item is an integer and the other is a list, promote the
  integer to a list and then compare as above.

  once you know an item is out of order, you are done. }

function comparelist(x1, x2 : plist) : integer;

var
   d  : integer;
   tx : plist;

begin
   { assume out of order }
   comparelist := +high(int64);

   while notnilorempty(x1) and notnilorempty(x2) do begin
      if hasvalue(x1) and hasvalue(x2) then begin
         d := getvalue(x1) - getvalue(x2);
         if d <> 0 then begin
            comparelist := d;
            exit;
         end;
         x1 := x1^.next; x2 := x2^.next;
         continue;
      end;
      if hasnested(x1) and hasnested(x2) then begin
         d := comparelist(getnested(x1), getnested(x2));
         if d <> 0 then begin
            comparelist := d;
            exit;
         end;
         x1 := x1^.next; x2 := x2^.next;
         continue;
      end;
      if hasvalue(x1) and hasnested(x2) then begin
         tx := listfromvalue(x1);
         d := comparelist(tx, getnested(x2));
         freelist(tx);
         if d <> 0 then begin
            comparelist := d;
            exit;
         end;
         x1 := x1^.next; x2 := x2^.next;
         continue;
      end;
      if hasnested(x1) and hasvalue(x2) then begin
         tx := listfromvalue(x2);
         d := comparelist(getnested(x1), tx);
         freelist(tx);
         if d <> 0 then begin
            comparelist := d;
            exit;
         end;
         x1 := x1^.next; x2 := x2^.next;
         continue;
      end;
      aochalt('invalid list structure! abort in compare');
      break;
   end;
   { if both lists are exhausted, 'equal', if left empty, in order, if right empty, out of order }
   if isnilorempty(x1) and isnilorempty(x2) then
      comparelist := 0
   else if isnilorempty(x1) then
      comparelist := -high(int64)
   else
      comparelist := +high(int64);
end;


{ read and parse into two arrays for parts one and two. part two wants
  the lists sorted, so since we're already doing a compare to check
  for pair ordering, lay each pair into the second array in order to
  save a few compare and swaps later on. }

procedure load;

var
   i, j    : integer;
   s, t, u : string;

begin
   aocrewind;
   i := 0; j := -1;
   while not aoceof do begin
      i := i + 1; j := j + 2;
      s := aocread;
      t := aocread;
      u := aocread;
      if u <> '' then
         aochalt('*** error in input ***');
      with pair[i] do begin
         left := aslist(s);
         right := aslist(t);
         inorder := comparelist(left, right);
         if inorder < 1 then begin        { a very minor optimization }
            packet[j] := left;
            packet[j+1] := right;
         end else begin
            packet[j] := right;
            packet[j+1] := left;
         end;
      end;
   end;
   numpairs := i;
   numpackets := i*2;
end;


{ a not so quick sort of the packets into ascending order }

procedure sort;

var
   i : integer;
   f : boolean;
   s : plist;

begin
   f := true;
   while f do begin
      f := false;
      for i := 1 to numpackets-1 do
         if comparelist(packet[i], packet[i+1]) > 0 then begin
            f := true;
            s := packet[i];
            packet[i] := packet[i+1];
            packet[i+1] := s;
         end;
   end;
end;


{ find a particular list value in the packets }

function indexof(x : plist) : integer;

var
   i : integer;

begin
   indexof := -1;
   for i := 1 to numpackets do
      if comparelist(packet[i], x) = 0 then begin
         indexof := i;
         break;
      end;
end;


{ mainline }

var
   i        : integer;
   partone  : integer;
   parttwo  : integer;
   pk1, pk2 : integer;

begin
   aocopen;
   init;
   load;
   partone := 0;
   for i := 1 to numpairs do begin
      if pair[i].inorder < 1 then
         partone := partone + i;
   end;
   packet[numpackets+1] := aslist('[[2]]');
   packet[numpackets+2] := aslist('[[6]]');
   numpackets := numpackets + 2;
   sort;
   pk1 := indexof(aslist('[[2]]'));
   pk2 := indexof(aslist('[[6]]'));
   parttwo := pk1 * pk2;
   writeln('part one answer : ', partone);
   writeln('part two answer : ', parttwo);
   aocclose;
end.

{ end solution.pas }

File: 2022/13/solution.f90
! working theory ... data read isn't being stored.
! solution.f90 -- 2022 day 13 -- distress signal -- troy brumley blametroi@gmail.com

! advent of code 2022 day 13 -- distress signal
!
! for part one, identify communication packets that are correctly ordered. the
! packets are lists and this looks rather lispish in lean, but we'll go with
! another language.

program solution

   use iso_fortran_env, only: int64
   use aocinput, only: open_aoc_input, read_aoc_input, close_aoc_input, max_aoc_reclen, rewind_aoc_input

   implicit none

   integer, parameter            :: AOCIN = 10
   integer(kind=int64)           :: part_one, part_two  ! results
   character(len=max_aoc_reclen) :: rec
   integer, parameter            :: MAXPAIRS = 500      ! observed

   ! the only way i've been able to get an algorithm that doesn't offend my
   ! sensibilities is to get closer and closer to a lisp list. i guess they knew
   ! what they were doing way back then. a list is at least two nodes. the last
   ! node is always empty. as the problem for advent of code is dealing with lists
   ! of integers, the value payload is just an integer.
   !
   ! pointers in fortran are a bit different from pascal or c. i think they are
   ! better semantically, more like a base register for a dsect from my assembly
   ! language days, but my muscle memory with pascal and c is sometimes causing
   ! errors. worth the struggle, but still a struggle.

   ! an atomic item, an integer for this problem.
   type atom_t
      logical               :: isnull
      integer               :: val
   end type atom_t

   ! a list of lists or atoms
   type list_t
      type(atom_t), pointer :: atom
      type(list_t), pointer :: nested
      type(list_t), pointer :: next
   end type list_t

   ! input is multiple paired lists. this stores the parsed values of
   ! each pair and notes if they are in proper order.

   type pair_t
      type(list_t), pointer :: left, right
      logical :: inorder
   end type pair_t

   ! the pairs are for part one, the packets are for part two. note that the
   ! vectors here contain structures with pointers, they are not pointers as
   ! they were in the original pascal version.

   type(pair_t)            :: pair(1:MAXPAIRS)
   type(list_t), target    :: packet(1:MAXPAIRS * 2)
   type(list_t), pointer   :: templist

   integer                 :: numpairs, numpackets, curr, pk1, pk2

   ! open and initialize.

   rec = ""
   call open_aoc_input(AOCIN)
   part_one = 0; part_two = 0
   call init

   ! load pairs and seed the packets, but packets will need sorting for
   ! part two.

   call load

   ! for part one, sum the indices of the in-order pairs

   do curr = 1, numpairs
      if (pair(curr) % inorder) part_one = part_one + curr
   end do

   ! for part two, insert two missing 'divider' packets, sort them with the
   ! other packets, and report the product of their indices as a key to the
   ! solution.

   templist => strlist("[[2]]")
   packet(numpackets + 1) = templist
   templist => strlist("[[6]]")
   packet(numpackets + 2) = templist
   numpackets = numpackets + 2

   call sort

   pk1 = indexof(strlist("[[2]]"))
   pk2 = indexof(strlist("[[6]]"))
   part_two = pk1 * pk2

   ! report and close

   print *
   print *, part_one, "part one"
   print *, part_two, "part two"
   print *

   call close_aoc_input(AOCIN)

contains

   ! make sure all pointers are unassociated.

   subroutine init
      implicit none
      integer :: i

      do i = 1, MAXPAIRS

         nullify (pair(i) % left)

         nullify (pair(i) % right)

         nullify (packet(i * 2 - 1) % atom)
         nullify (packet(i * 2 - 1) % nested)
         nullify (packet(i * 2 - 1) % next)

         nullify (packet(i * 2) % atom)
         nullify (packet(i * 2) % nested)
         nullify (packet(i * 2) % next)

      end do

      numpairs = 0
      numpackets = 0
   end subroutine init

   ! read and parse into two arrays for parts one and two. part two wants the
   ! lists sorted, so since we're already doing a compare to check for pair
   ! ordering, lay each pair into the second array in order to (hopefully)
   ! save a few compare and swaps later on.

   subroutine load
      implicit none
      integer :: i, j
      character(len=max_aoc_reclen) :: s, t

      call rewind_aoc_input(AOCIN)

      ! input is at least two lines, with a third blank separator line if more
      ! pairs follow. i'm being needlessly cute here and allocating "just big
      ! enough" buffers to hold each line of input as i read them just to get
      ! comfortable with the fortran way of doing allocations. it's not really
      ! needed.
      !
      ! pairs subscript from one and advance by one (i), while packets subscript
      ! from one and advance by two (j) for each logical input.

      i = 0; j = -1

      readloop: do while (read_aoc_input(AOCIN, rec))
         i = i + 1; j = j + 2

         ! copy left to buffer. i know i could keep track of the size myself and
         ! deallocate/reallocate only when the incoming record is longer than the
         ! current buffer, but i wonder if there's a way to interrogate the size
         ! of the allocated area.

         s = rec(1:len_trim(rec))

         ! read right and copy to another buffer.

         if (.not. read_aoc_input(AOCIN, rec)) then
            print *, "*** error *** unexpected end of file"
            stop 1
         end if
         t = rec(1:len_trim(rec))

         ! advance past separator line. an end of file here is ok and will
         ! be caught by the while test above.

         if (read_aoc_input(AOCIN, rec)) then
            if (len_trim(rec) /= 0) then
               print *, "*** error *** error in input"
               stop 1
            end if
         end if

         ! parse the pair

         pair(i) % left => strlist(s)
         pair(i) % right => strlist(t)

         ! while we're here, note if they are in order for part one

         pair(i) % inorder = listcompare(pair(i) % left, pair(i) % right) < 1

         ! copy to packets for part two, order them on the way in even though it
         ! probably won't help the later sort much note: this copies the data of
         ! the root list_t, not a pointer to it, from the pair set. the anchor
         ! arrays come with the head of each list already there, and then we
         ! chain off them. as there are no backreferences, this copy shouldn't
         ! cause any problems.

         if (pair(i) % inorder) then
            packet(j) = pair(i) % left
            packet(j + 1) = pair(i) % right
         else
            packet(j) = pair(i) % right
            packet(j + 1) = pair(i) % left
         end if

      end do readloop

      numpairs = i
      numpackets = i * 2

   end subroutine load

   ! gfortran print *,int produces a lot of leading blanks. lfortran does not.
   ! intstr gives the integer as a string with only one leading and one
   ! trailing blank.

   function intstr(n) result(res)
      implicit none
      integer, intent(in)           :: n
      character(len=:), allocatable :: res
      character(len=32)             :: buf

      buf = ""
      write (buf, *) n
      allocate (character(len=len_trim(adjustl(buf))) :: res)
      res = trim(adjustl(buf))
   end function intstr

   ! fortran's displayable logical is T or F. let's get wordy.

   ! function boolstr(b) result(res)
   !    implicit none
   !    logical, intent(in)           :: b
   !    character(len=:), allocatable :: res
   !    if (b) then
   !       allocate (character(len=4) :: res)
   !       res = "True"
   !    else
   !       allocate (character(len=5) :: res)
   !       res = "False"
   !    end if
   ! end function boolstr

   ! compare two lists using the rules for part one and return via the negative
   ! zero positive convention.
   !
   ! compare the lists item by item, with the following behaviors for nested
   ! lists:
   !
   ! * a longer list comes after an otherwise equivalent list
   ! * if one item is an integer and the other is a list, promote the integer
   !   to a list and then compare as above.
   !
   ! once you know an item is out of order, you are done.

   recursive function listcompare(xin, yin) result(res)
      implicit none
      type(list_t), pointer, intent(in) :: xin, yin
      integer                           :: res
      integer                           :: d
      type(list_t), pointer             :: xs, ys, tx, ty

      ! assume they are out of order
      res = +huge(d)

      ! we need to chase through the links, so get local
      ! updatable copies of the list anchors
      xs => xin
      ys => yin

      ! item/position by item/position

      compareloop: do while (notnilorempty(xs) .and. notnilorempty(ys))

         ! easy case one, two atoms.

         if (hasvalue(xs) .and. hasvalue(ys)) then
            d = getvalue(xs) - getvalue(ys)
            if (d /= 0) then
               res = d
               return
            end if
            xs => xs % next; ys => ys % next
            cycle compareloop
         end if

         ! easy case two, two nested lists.

         if (hasnested(xs) .and. hasnested(ys)) then
            d = listcompare(getnested(xs), getnested(ys))
            if (d /= 0) then
               res = d
               return
            end if
            xs => xs % next; ys => ys % next
            cycle compareloop
         end if

         ! less easy case one, atom to list. here we cast the atom to a list and
         ! compare it to the other list, being sure to free the temporary list
         ! when the compare completes.

         if (hasvalue(xs) .and. hasnested(ys)) then
            tx => listfromvalue(xs)
            d = listcompare(tx, getnested(ys))
            call freelist(tx)
            if (d /= 0) then
               res = d
               return
            end if
            xs => xs % next; ys => ys % next
            cycle compareloop
         end if

         ! same song, other foot.

         if (hasnested(xs) .and. hasvalue(ys)) then
            ty => listfromvalue(ys)
            d = listcompare(getnested(xs), ty)
            call freelist(ty)
            if (d /= 0) then
               res = d
               return
            end if
            xs => xs % next; ys => ys % next
            cycle compareloop
         end if

         ! should not occur

         print *, "*** error *** invalid list structure, aborting in listcompare"
         stop 1

      end do compareloop

      ! if both lists are exhausted, equal. if left empty, in order. if right
      ! empty, out of order.

      if (isnilorempty(xs) .and. isnilorempty(ys)) then
         res = 0
      else if (isnilorempty(xs)) then
         res = -huge(d)
      else
         res = huge(d)
      end if
   end function listcompare

   ! list node accessors and predicates. i prefer an api instead of expposing
   ! the structure/link paths in higher level code.

   ! a list can contain another list. xs is marked inout, but we only update
   ! its members, not its value.

   subroutine setnested(xs, ys)
      implicit none
      type(list_t), pointer, intent(inout) :: xs
      type(list_t), pointer, intent(in) :: ys

      if (associated(xs % atom)) then
         print *, "*** error *** assigning nested to an atomic node"
         stop 1
      end if
      xs % nested => ys
   end subroutine setnested

   function getnested(xs) result(res)
      implicit none
      type(list_t), pointer :: res
      type(list_t), pointer, intent(in) :: xs

      res => xs % nested
   end function getnested

   ! a list can contain an atomic value, in this application, an integer. the
   ! value could be stored directly, but if this where a real application and
   ! more atomic types were added, it would make more sense and be a better use
   ! of storage to keep the list nodes as small as possible, with the varying
   ! payload sizes moved to separate structures.

   subroutine setvalue(xs, n)
      implicit none
      type(list_t), pointer, intent(inout) :: xs
      type(integer), intent(in) :: n

      if (associated(xs % nested)) then
         print *, "*** error *** assigning value to nesting node"
         stop 1
      end if

      ! allocate storage for the atom if needed. if there is already an atom, overwrite
      ! its value.

      if (.not. associated(xs % atom)) allocate (xs % atom)
      xs % atom % isnull = .false.
      xs % atom % val = n
   end subroutine setvalue

   function getvalue(xs) result(res)
      implicit none
      integer :: res
      type(list_t), pointer, intent(in) :: xs

      res = xs % atom % val
   end function getvalue

   ! list nodes chain togeter to form the actual list
   ! subroutine setnext(xs, ys)
   !    implicit none
   !    type(list_t), pointer, intent(inout) :: xs
   !    type(list_t), pointer, intent(in) :: ys
   !    xs%next => ys
   ! end subroutine setnext

   ! expand by null: creates a new empty list node, links it to xs,
   ! and advances xs.

   subroutine addnewempty(xs)
      implicit none

      type(list_t), pointer, intent(inout) :: xs

      allocate (xs % next)

      xs => xs % next
   end subroutine addnewempty

   ! various list node predicates
   ! i prefer functions to having all the root%member%member%member visible topside

   function isempty(x) result(res)
      implicit none
      logical :: res
      type(list_t), pointer, intent(in) :: x

      res = .not. (associated(x % atom) .or. associated(x % nested))
   end function isempty

   function hasvalue(x) result(res)
      implicit none
      logical :: res
      type(list_t), pointer, intent(in) :: x

      res = associated(x % atom)
   end function hasvalue

   function hasnested(x) result(res)
      implicit none
      logical :: res
      type(list_t), pointer, intent(in) :: x

      res = associated(x % nested)
   end function hasnested

   ! function hasnext(x) result(res)
   !    implicit none
   !    logical :: res
   !    type(list_t), pointer, intent(in) :: x
   !    res = associated(x%next)
   ! end function hasnext

   function isnilorempty(x) result(res)
      implicit none
      logical :: res
      type(list_t), pointer, intent(in) :: x

      res = .not. associated(x)
      if (.not. res) res = isempty(x)
   end function isnilorempty

   function notnilorempty(x) result(res)
      implicit none
      logical :: res
      type(list_t), pointer, intent(in) :: x

      res = .not. isnilorempty(x)
   end function notnilorempty

   ! release the list any linked atom values and sub lists, clearing any storage
   ! before it is released.

   recursive subroutine freelist(xs)
      implicit none
      type(list_t), pointer, intent(inout) :: xs
      type(list_t), pointer :: nx

      assocloop: do while (associated(xs))
         nx => xs % next

         ! free any children
         if (hasnested(xs)) then
            call freelist(xs % nested)
            nullify (xs % nested)
         end if

         ! free any atom
         if (associated(xs % atom)) then
            xs % atom % isnull = .true.
            xs % atom % val = 0
            deallocate (xs % atom)
         end if

         ! release
         nullify (xs % next)
         deallocate (xs)

         ! advance
         xs => nx
      end do assocloop

   end subroutine freelist

   ! create a new list from a value
   function listfromvalue(x) result(res)
      implicit none
      type(list_t), pointer :: res
      type(list_t), pointer, intent(in) :: x

      allocate (res)
      if (hasvalue(x)) call setvalue(res, getvalue(x))
   end function listfromvalue

   ! recursive helper for strlist. should be called with pos pointing at a list
   ! open bracket [ and will return as each list close bracket ] is reached. if
   ! everything is nested properly, it works. there is no attempt at input
   ! validation.

   recursive function strlistr(pos, str) result(res)
      implicit none
      type(list_t), pointer :: res
      character(len=*), intent(in) :: str
      integer, intent(inout) :: pos

      ! xs and ys are modifiable work pointers
      type(list_t), pointer :: xs, ys
      integer :: n

      ! our result is the first item, subsequent items are linked
      ! following. xs is bumped as needed.

      allocate (xs)
      res => xs

      ! guard clauses and special cases

      ! should not happen, just return an empty list
      if (len_trim(str) < 2) then
         return
      end if

      ! an actual empty list is legal but won't happen until i clean
      ! up the loader
      if (str(pos:pos) == "[" .and. str(pos + 1:pos + 1) == "]") then
         pos = pos + 2
         return
      end if

      ! this better be a properly bracketed list
      n = len_trim(str)
      if (str(pos:pos) /= "[" .or. str(n:n) /= "]") then
         print *, "*** error *** misplaced open bracket, skipping ahead one from", intstr(pos)
         pos = pos + 1
         return
      end if

      ! consume the opening [

      pos = pos + 1

      ! and scan through the list

      scanner: do while (pos <= len_trim(str))

         ! commas and spaces are all whitespace

         if (str(pos:pos) == " " .or. str(pos:pos) == ",") then
            pos = pos + 1
            cycle scanner
         end if

         ! end of the current (sub) list? if so, bump and return

         if (str(pos:pos) == "]") then
            pos = pos + 1
            return ! <---------- return from function
         end if

         ! start of a nested (sub) list? if so, call recursively and tack it on to
         ! the list i'm working on.

         if (str(pos:pos) == "[") then
            ys => strlistr(pos, str)
            call setnested(xs, ys)
            call addnewempty(xs)
            cycle scanner
         end if

         ! add atom to list, only ints are supported

         if (str(pos:pos) >= "0" .and. (str(pos:pos) <= "9")) then
            read (str(pos:len_trim(str)), *) n
            call setvalue(xs, n)
            call addnewempty(xs)
            pos = pos + max(1, scan(str(pos:len_trim(str)), " ,[]"))
            cycle scanner
         end if

         ! input format error, report but continue

         print *, "*** error *** at position ", intstr(pos), " in '", trim(str), "'"
         pos = pos + 1

      end do scanner

      print *, "*** error *** list parse failure or unbalanced brackets"
      stop 1
   end function strlistr

   ! from a list of integers and nested lists of integers in string form,
   ! create a navigable set of nodes. the ugly details are in a recursive
   ! helper function.

   function strlist(str) result(res)
      implicit none
      character(len=*), intent(in)  :: str
      type(list_t), pointer         :: res
      character(len=:), allocatable :: sanitized
      integer                       :: pos, i, j, k

      ! read() in strlistr wants a space or comma after integers, not a ] so pad
      ! [ and ] with spaces

      ! how many brackets are there?

      k = 0
      do i = 1, len_trim(str)
         if (str(i:i) == "[") k = k + 1
         if (str(i:i) == "]") k = k + 1
      end do

      ! get enough room for the padding

      allocate (character(len=len_trim(str) + k*3) :: sanitized)
      sanitized = repeat(" ", len_trim(str) + k * 3)

      ! and pad each bracket with a space on either side.

      j = 1
      sanitizer: do i = 1, len_trim(str)
         if (str(i:i) == "[") then
            sanitized(j:j + 2) = " [ "
            j = j + 3
            cycle sanitizer
         end if
         if (str(i:i) == "]") then
            sanitized(j:j + 2) = " ] "
            j = j + 3
            cycle sanitizer
         end if
         sanitized(j:j) = str(i:i)
         j = j + 1
      end do sanitizer

      ! call recursive helper to parse into a list

      pos = 1
      res => strlistr(pos, adjustl(sanitized))

      ! i'm not sure this is needed since we're going out of scope, but
      ! my habit is to release dynamic memory while running.

      deallocate (sanitized)

   end function strlist

   ! given a list of nodes, return a string representation.
   ! overallocating strings just to get this working quickly.

   recursive function liststr(xin) result(res)
      implicit none
      type(list_t), pointer, intent(in) :: xin
      character(len=4096)               :: res
      type(list_t), pointer             :: xs
      integer                           :: i

      ! copy to local so we can step through the list
      xs => xin

      res = "["

      do while (notnilorempty(xs))

         if (hasvalue(xs)) then
            ! this leaves a dangling comma
            res = trim(res)//trim(intstr(getvalue(xs)))//","
            xs => xs % next
            cycle
         end if

         if (hasnested(xs)) then
            res = trim(res)//trim(liststr(getnested(xs)))
            xs => xs % next
            cycle
         end if

         ! i don't think we should get here, but ...

         print *, "*** error *** listtstr empty node, skipping"
         xs => xs % next

      end do

      ! if there's a dangling comma, remove it

      i = len_trim(res)
      if (res(i:i) == ",") res(i:i) = " "

      ! close the bracket and we're done
      res = trim(res)//"]"

   end function liststr

   ! a not so quick exchange sort of the packets into ascending order

   subroutine sort
      implicit none
      integer               :: i
      logical               :: f
      type(list_t), pointer :: xs, ys
      type(list_t)          :: swap

      f = .true.
      outer: do while (f)
         f = .false.
         inner: do i = 1, numpackets - 1
            xs => packet(i); ys => packet(i + 1)
            if (listcompare(xs, ys) > 0) then
               f = .true.
               swap = packet(i)
               packet(i) = packet(i + 1)
               packet(i + 1) = swap
            end if
         end do inner
      end do outer
   end subroutine sort

   ! find a matching list value in the packets and return it index. if no match
   ! is found -1 is returned. while the packets array is sorted for this problem,
   ! this search does not count on the items being in order.

   function indexof(xin) result(res)
      implicit none
      type(list_t), pointer, intent(in) :: xin
      integer                           :: res
      type(list_t), pointer             :: ys
      integer                           :: i

      res = -1
      do i = 1, numpackets
         ys => packet(i)
         if (listcompare(xin, ys) == 0) then
            res = i
            return
         end if
      end do
   end function indexof

end program solution
